File: 1032-satisfiability-of-equality-equations/satisfiability-of-equality-equations.py
class Solution:
    def __init__(self):
        self.parent = []
        self.rank = []

    def find(self, i: int) -> int:
        if self.parent[i] != i:
            self.parent[i] = self.find(self.parent[i])
        return self.parent[i]

    def Union(self, x: int, y: int) -> None:
        parent_x = self.find(x)
        parent_y = self.find(y)
        if parent_x != parent_y:
            if self.rank[parent_x] > self.rank[parent_y]:
                self.parent[parent_y] = parent_x
            elif self.rank[parent_x] < self.rank[parent_y]:
                self.parent[parent_x] = parent_y
            else:
                self.parent[parent_x] = parent_y
                self.rank[parent_y] += 1

    def equationsPossible(self, equations: List[str]) -> bool:
        self.parent = list(range(26))
        self.rank = [1] * 26

        for s in equations:
            if s[1] == "=":
                self.Union(ord(s[0]) - ord("a"), ord(s[3]) - ord("a"))

        for s in equations:
            if s[1] == "!":
                if self.find(ord(s[0]) - ord("a")) == self.find(ord(s[3]) - ord("a")):
                    return False
        return True

File: 1032-satisfiability-of-equality-equations/satisfiability-of-equality-equations.cpp
class Solution {
public:
    vector<int> parent;
    vector<int> rank;
    int find(int i) {
        if (parent[i] != i) {
            parent[i] = find(parent[i]);
        }
        return parent[i];
    }
    void Union(int x, int y) {
        int parent_x = find(x);
        int parent_y = find(y);
        if (parent_x != parent_y) {
            if (rank[parent_x] > rank[parent_y]) {
                parent[parent_y] = parent_x;
            } else if (rank[parent_x] < rank[parent_y]) {
                parent[parent_x] = parent_y;
            } else {
                parent[parent_x] = parent_y;
                rank[parent_y]++;
            }
        }
    }
    bool equationsPossible(vector<string>& equations) {
        parent.resize(26);
        rank.resize(26);
        for (int i = 0; i < 26; i++) {
            parent[i] = i;
            rank[i] = 1;
        }
        for (string &s : equations) {
            if (s[1] == '=') {
                Union(s[0] - 'a', s[3] - 'a');
            }
        }
        for (string &s : equations) {
            if (s[1] == '!') {
                if (find(s[0] - 'a') == find(s[3] - 'a'))
                    return false;
            }
        }
        return true;
    }
};